File: Programming/Java/Container With Most Water/containerWithMostWater.java

import java.util.*;

public class containerWithMostWater  
{
	
	// method to calculate area of container with max water
	public static int calculateArea(int[] arr, int n)
	{
    	int left = 0;			// co-ordinate of left wall
		int right = n-1;		// co-ordinate of right wall
		int maxArea=0;			// store of area of conatainer with max water
		
		// lopoing till left wall crosses right wall
		while(left < right)
		{
			
			// height of container will be height of smaller wall
			// base length will be how far they are apart i.e., right - left
			
			int area = Math.min(arr[left],arr[right])*(right-left);
			
			
			// storing the max calculated area of container
			maxArea = Math.max(maxArea, area);
			
			// moving towards right to find left wall of higher height 
			if(arr[left] < arr[right])
				left++;
			
			// moving towards left to find right wall of higher height
			else
				right--;
			
		}
		return maxArea;
  	}

	// driver program
	public static void main(String[] args)
	{
		Scanner sc = new Scanner (System.in);
		
		int n = sc.nextInt();			// number of walls
		int arr[] = new int[n];			// for storing height of walls
		
		for(int i = 0;i < n; i++)
			arr[i] = sc.nextInt();		// input height of walls
		
		
		System.out.println(calculateArea(arr, n));
	}
}

File: Programming/C/ContainerWithMostWater/ContainerWithMostWater.c
#include<stdio.h>
int mostWater(int arr[], int length);
int max(int num1, int num2);
int min(int num1, int num2);
int main() {
    int arr[] = {1,8,6,2,5,4,8,3,7};
    int length = sizeof(arr)/sizeof(arr[0]);
    int res = mostWater(arr,length);
    printf("%d" , res);
    return 0;
}
int mostWater(int arr[], int length) {
    int l = 0, r = length-1;
    int max_area = 0;
    while (l < r) 
    {
        max_area = max(max_area,(min(arr[l],arr[r])*(r-l)));
        if(arr[r] > arr[l]) l++;
        else r--;
    }
    return max_area;
    
}
int max(int num1, int num2)
{
    return (num1 > num2 ) ? num1 : num2;
}
int min(int num1, int num2) 
{
    return (num1 > num2 ) ? num2 : num1;
}